### PR TITLE
chore: delete all stacks on close

### PR DESCRIPTION
**Pull request by thantos**
Stacks stay alive after the PR is closed, this PR will delete all stacks related to the PR on PR close.

### CODE REVIEW

File: test/cleanup.runtime.test.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+          (resource) =>
+            $AWS.SDK.CloudFormation.deleteStack(
+              { StackName: resource.ResourceARN! },
+              {
+                iam: {
+                  resources: [stackArnPrefix],
+                },
+              }
+            )
+        );
```
Comment (sam-goodwin) [inline test/cleanup.runtime.test.ts:70]: Super nice - love it.

File: test/cleanup.runtime.test.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+      // set lambda url in ssm parameter
+      // grant access to ?? to call ssm and lambda url
+      // const stackPartition = new Function(
+      //   scope,
+      //   "stackPartition",
+      //   async (stackArns: string[]) => ({
+      //     cleanUpStack: stackArns.find((arn) => arn.includes("/cleanUp/")),
+      //     otherStacks: stackArns.filter((arn) => !arn.includes("/cleanUp/")),
+      //   })
+      // );
```
Comment (sam-goodwin) [inline test/cleanup.runtime.test.ts:32]: is this no longer needed? Found a functionless way to do it?

File: test/cleanup.runtime.test.ts [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        parameterName: `/functionlessTestDeleter/${runtimeTestExecutionContext.stackTag}/deleteUrl`,
+      });
+
+      return {
+        outputs: {},
+      };
+    },
+    async () => {
+      expect(true).toBeTruthy();
+    }
```
Comment (sam-goodwin) [inline test/cleanup.runtime.test.ts:94]: Is this meant to do something or are you just using the harness?

File: test/cleanup.runtime.test.ts [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+                  Key: STACK_TAG_KEY,
+                  Values: [runtimeTestExecutionContext.stackTag],
+                },
+              ],
+              ResourceTypeFilters: ["cloudformation:stack"],
+            },
+            {
+              iam: {
+                resources: ["*"],
+              },
```
[the remark this answers is not in the document]
Comment (thantos) [reply test/cleanup.runtime.test.ts:55]: yes, stacks are computed and (currently) have no shared prefix. We could use the tags though in conditions...

File: test/cleanup.runtime.test.ts (hunk 1 above)
Comment (thantos) [reply test/cleanup.runtime.test.ts:70]: right? nearly effortless with our library... (except when I tried to use (`string.includes` #479)  and (`string.split` #468)

File: test/cleanup.runtime.test.ts (hunk 2 above)
Comment (thantos) [reply test/cleanup.runtime.test.ts:32]: I was going to wait to delete the cleanup stack, but... found that step functions waits for running jobs to finish anyways, so it s safe to just delete them all.

Forgot to delete the extra code.

File: test/cleanup.runtime.test.ts (hunk 3 above)
Comment (thantos) [reply test/cleanup.runtime.test.ts:94]: just using the harness :-)

File: test/cleanup.runtime.test.ts (hunk 1 above)
Comment (sam-goodwin) [reply test/cleanup.runtime.test.ts:70]: Haha you "played yourself" ;)

File: test/cleanup.runtime.test.ts (hunk 3 above)
Comment (sam-goodwin) [reply test/cleanup.runtime.test.ts:94]: Harness beginning to feel like more than a test utility?

**Comment by thantos** [comment]
Getting creative trying to test the on close script....

File: .projenrc.ts [hunk 5]
```diff
@@ -198,6 +201,40 @@ packageJson.addOverride("lint-staged", {
   "*.{tsx,jsx,ts,js,json,md,css}": ["eslint --fix"],
 });
 
+const closeWorkflow = project.github?.addWorkflow("close");
+closeWorkflow?.on({
+  pullRequest: {
+    types: ["closed"],
```
Comment (sam-goodwin) [inline .projenrc.ts:207]: Does this include merges and cancels?

**Review by sam-goodwin — APPROVED** [review]

File: .projenrc.ts (hunk 5 above)
Comment (thantos) [reply .projenrc.ts:207]: Yeah. Merge is just a flag on closed.

**Comment by thantos** [comment]
well... forgot to remove the `needs` on the job so this broke... sigh...

> The workflow is not valid. .github/workflows/close.yml (Line: 9, Col: 3): The workflow must contain at least one job with no dependencies.

https://github.com/functionless/functionless/actions/runs/3053133159

**Pull request merged**